### ISSUE TITLE
[Bug] Logical operators are not evaluated lazily

### ISSUE THREAD

**Issue by slisaasquatch** [IBM/JSONata4Java#177]
Here's a simple code sample that demonstrates the issue:

```java
	public static void main(String[] args) throws Exception {
		final ObjectNode ctx = JsonNodeFactory.instance.objectNode()
				.put("email", (String) null);
		System.out.println("ctx: " + ctx);
		final JsonNode result = Expressions.parse("email and $contains(email, \"foo\")").evaluate(ctx);
		System.out.println(result);
	}
```

The code causes the following error

```
Exception in thread "main" com.api.jsonata4java.expressions.EvaluateException: Argument 1 of function $contains does not match function signature
	at com.api.jsonata4java.expressions.Expressions.evaluate(Expressions.java:194)
```

The same example [here at https://try.jsonata.org/YjjQK644Q](https://try.jsonata.org/YjjQK644Q) simply returns `false`.

**Comment by wnm3** [comment]
I've asked this [question in the jsonata channel](https://jsonata.slack.com/archives/C2L3MR5EH/p1639773819061000) to determine if you've uncovered an exception in jsonata:
Given an object like { "email": null } the expression:
email and $contains(email, "foo") returns false
but both:
$contains(email, "foo")
$contains(email, "foo") and email
return the exception: Argument 1 of function "contains" does not match function signature
This seems somewhat arbitrary. Do logic expressions treat exceptions as false or should they expose an error in their operand.
Note: email or $contains(email, "foo") as well as $not($contains(email,"foo")) also throws the exception...
This was reported as a bug in JSONata4Java but before I "fix" it I wondered if this was a bug here ;^) -- e.g., the and expression should also report the exception condition.

**Comment by slisaasquatch** [comment]
Personally I think the behaviour of jsonata-js makes more sense. The logical operator is not exactly treating exceptions as false. It is skipping the evaluation of the `$contains` because it doesn't need to. This is how most programming languages behave anyway. Take the following simple Java code for example:

```java
	public static void main(String[] args) {
		System.out.println(true || foo() > 0);
	}

	static int foo() {
		throw new RuntimeException();
	}
```

It will print `true` instead of throwing an exception. It is also how most other programming languages work.

**Comment by wnm3** [comment]
I don't disagree with you but I'm trying to determine where to drop out of the expression evaluation. I believe you are proposing if the LHS of an or condition is true, go ahead and return true. But another side of me would want the author to know there is a potential bug in their expression should they ever activate that portion of the code as they'll believe it has worked forever so why did adding another condition break things (e.g., changing if (true || bad_statement) to become if ((x && true) || bad_statement ). Let me see how the discussion pans out on jsonata.org's slack channel.

**Comment by wnm3** [comment]
I went ahead and released version 1.6.1 (should be in maven in a couple hours) with the patch to return true if the LHR for or is true, and return false if the LHR for and is false.

Please test and close this issue if you are satisfied. Thank you.

**Comment by slisaasquatch** [comment]
I can confirm that that fixed the issue. Thanks for the extremely quick turnaround!
